### PR TITLE
Throttle client

### PR DESCRIPTION
**Pull request by sjezewski**
Applies a buffer of semaphores to all `PutFile` and `GetFile` calls. The size is determined by the argument passed to the client creation functions `NewFromAddress()` or `NewMetricsClientFromAddress()`

This allows us to buffer put/getfiles:

- during jobshim download phase
- during jobshim upload phase
- during pachctl putfile / getfile

Additionally, fixed a bug w the semaphores in pachd's `PutFile` when putting a list of URLs.

There are no tests for these features because it requires load testing. Tests for each of these cases will go into the load testing harness.

### CODE REVIEW

File: src/client/client.go [hunk 1]
```diff
@@ -42,11 +42,15 @@ type APIClient struct {
 	cancel            func()
 	reportUserMetrics bool
 	metricsPrefix     string
+	streamSemaphores  chan struct{}
```
Comment (jdoliner) [inline src/client/client.go]: Minor knit, don't think this should be plural since it's just a single semaphore.

File: src/client/client.go [hunk 2]
```diff
@@ -63,9 +67,10 @@ func NewMetricsClientFromAddress(addr string, metrics bool, prefix string) (*API
 }
 
 // NewFromAddress constructs a new APIClient for the server at addr.
-func NewFromAddress(addr string) (*APIClient, error) {
+func NewFromAddress(addr string, maxConcurrentStreams uint) (*APIClient, error) {
```
Comment (jdoliner) [inline src/client/client.go]: Changing this signature seems like it's likely to break some existing user code. Could we use default here and have a different method for creating one with a non default number of streams?

File: src/client/pfs.go [hunk 3]
```diff
@@ -447,6 +447,10 @@ func (c APIClient) PutFileWriter(repoName string, commitID string, path string,
 
 // PutFile writes a file to PFS from a reader.
 func (c APIClient) PutFile(repoName string, commitID string, path string, reader io.Reader) (_ int, retErr error) {
+	c.streamSemaphores <- struct{}{}
+	defer func() {
+		<-c.streamSemaphores
+	}()
```
Comment (jdoliner) [inline src/client/pfs.go]: Minor stylistic thing, I generally like to make these into 1 liners to conserve vertical space.
Gofmt doesn't mind it either.

File: src/client/client.go (hunk 1 above)
Comment (sjezewski) [inline src/client/client.go]: Cool. Updated this

File: src/client/client.go (hunk 2 above)
Comment (sjezewski) [inline src/client/client.go]: Good point! Updated this and added another 'sugar' constructor so less code changes

File: src/client/pfs.go (hunk 3 above)
Comment (sjezewski) [inline src/client/pfs.go]: Ok, done

**Pull request merged**